src/test/java/com/cloudbees/plugins/credentials/CredentialsInPipelineTest.java

```diff
@@ -0,0 +1,702 @@
+                "KeyStoreSource kss = ((CertificateCredentialsImpl) credential).getKeyStoreSource();\n" +
+                "echo \"KSS: \" + kss.toString()\n" +
+                "byte[] kssb = kss.getKeyStoreBytes();\n" +
+                "echo \"KSS bytes (len): \" + kssb.length\n" +
+                "String keyValue = keyStore.getKey(alias, password.toCharArray()).getEncoded().encodeBase64().toString()\n" +
+                "echo  \"-----BEGIN PRIVATE KEY-----\"\n" +
+                "echo keyValue\n" +
+                "echo \"-----END PRIVATE KEY-----\"\n" +
+                "\n" +
+                "echo \"CRED-SNAP ON " + runnerTag + ":\"\n" +
+                "echo credentialSnap.toString()\n" +
+                "KeyStore keyStoreSnap = credentialSnap.getKeyStore();\n" +
+                "KeyStoreSource kssSnap = ((CertificateCredentialsImpl) credentialSnap).getKeyStoreSource();\n" +
+                "echo \"KSS-SNAP: \" + kssSnap.toString()\n" +
+                "byte[] kssbSnap = kssSnap.getKeyStoreBytes();\n" +
+                "echo \"KSS-SNAP bytes (len): \" + kssbSnap.length\n" +
+                "String keyValueSnap = keyStoreSnap.getKey(alias, password.toCharArray()).getEncoded().encodeBase64().toString()\n" +
+                "echo  \"-----BEGIN PRIVATE KEY-----\"\n" +
+                "echo keyValueSnap\n" +
+                "echo \"-----END PRIVATE KEY-----\"\n" +
+                "\n";
+    }
+
+    @Test
+    @Issue("JENKINS-70101")
+    public void testCertKeyStoreReadableOnController() throws Exception {
+        // Check that credentials are usable with pipeline script
+        // running without a node{}
+        prepareUploadedKeystore();
+
+        // Configure the build to use the credential
+        WorkflowJob proj = r.jenkins.createProject(WorkflowJob.class, "proj");
+        String script = cpsScriptCredentialTestImports() +
+                cpsScriptCertCredentialTestScriptedPipeline("CONTROLLER BUILT-IN");
+        proj.setDefinition(new CpsFlowDefinition(script, false));
+
+        // Execute the build
+        WorkflowRun run = proj.scheduleBuild2(0).get();
+        if (verbosePipelines) System.out.println(getLogAsStringPlaintext(run));
+
+        // Check expectations
+        r.assertBuildStatus(Result.SUCCESS, run);
+        // Got to the end?
+        r.assertLogContains("KSS-SNAP bytes", run);
+    }
+
+    @Test
+    @Issue("JENKINS-70101")
+    public void testCertKeyStoreReadableOnNodeLocal() throws Exception {
+        // Check that credentials are usable with pipeline script
+        // running on a node{} (provided by the controller)
+        prepareUploadedKeystore();
+
+        // Configure the build to use the credential
+        WorkflowJob proj = r.jenkins.createProject(WorkflowJob.class, "proj");
+        String script = cpsScriptCredentialTestImports() +
+                "node {\n" +
+                cpsScriptCertCredentialTestScriptedPipeline("CONTROLLER NODE") +
+                "}\n";
```

Comment:
The Groovy code in a Pipeline always executes on the controller. Wrapping everything in `node` does not actually change the behavior related to the issue you are seeing - this test `testCertKeyStoreReadableOnController`, and  `testCertKeyStoreReadableOnNodeRemote` should all have the same behavior as far as the credential is concerned regardless of your changes (unless I missed something in the Pipeline script that actually makes use of the agent). To actually make use of the `FilePath` or `Launcher` provided by the `node` you would need to write a `Step`.

Reply:
In my case it was more to rule out that nothing pops up with plain code and an unnamed node. It certainly did differ (original problem) for "named node" (or agent with label in declarative), with part of the plugin handling running on the remote agent.

Reply:
Yes, but in these three test cases, the nodes are not actually used by the credential-related code in the Pipelines. Regardless, I would replace your new tests with the minimal test from https://github.com/jenkinsci/credentials-plugin/commit/4ee16f3393421e6dae85d42445e56beff8ff9541.